pkg/chains/formats/intotoite6/extract/extract_test.go

```diff
@@ -0,0 +1,284 @@
+						{Name: "invalidDigestSHA_IMAGE_URL", Value: *v1beta1.NewStructuredValues(artifactURL2)},
+						{Name: "invalidDigestSHA_ARTIFACT_DIGEST", Value: *v1beta1.NewStructuredValues("sha256:x123")},
+						{Name: "invalidDigestSHA_ARTIFACT_URI", Value: *v1beta1.NewStructuredValues(artifactURL2)},
+						// missing digest algorithm name
+						{Name: "missingDigestAlgo_IMAGE_DIGEST", Value: *v1beta1.NewStructuredValues(artifactDigest2)},
+						{Name: "missingDigestAlgo_IMAGE_URL", Value: *v1beta1.NewStructuredValues(artifactURL2)},
+						{Name: "missingDigestAlgo_ARTIFACT_DIGEST", Value: *v1beta1.NewStructuredValues(artifactDigest2)},
+						{Name: "missingDigestAlgo_ARTIFACT_URI", Value: *v1beta1.NewStructuredValues(artifactURL2)},
+						// missing url value completely
+						{Name: "missingURLVal_IMAGE_DIGEST", Value: *v1beta1.NewStructuredValues(artifactDigest2)},
+						{Name: "missingURLVal_IMAGE_URL", Value: *v1beta1.NewStructuredValues("")},
+						{Name: "missingURLVal_ARTIFACT_DIGEST", Value: *v1beta1.NewStructuredValues(artifactDigest2)},
+						{Name: "missingURLVal_ARTIFACT_URI", Value: *v1beta1.NewStructuredValues("")},
+					},
+				},
+			},
+		},
+	)
+
+	prValid = objects.NewPipelineRunObject(
+		&v1beta1.PipelineRun{
+			Status: v1beta1.PipelineRunStatus{
+				PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+					PipelineResults: []v1beta1.PipelineRunResult{
+						{Name: "y_IMAGE_DIGEST", Value: *v1beta1.NewStructuredValues("sha256:" + artifactDigest1)},
+						{Name: "y_IMAGE_URL", Value: *v1beta1.NewStructuredValues(artifactURL1)},
+						{Name: "x_ARTIFACT_DIGEST", Value: *v1beta1.NewStructuredValues("sha256:" + artifactDigest2)},
+						{Name: "x_ARTIFACT_URI", Value: *v1beta1.NewStructuredValues(artifactURL2)},
+					},
+				},
+			},
+		},
+	)
+
+	prInvalidTypeHintingName = objects.NewPipelineRunObject(
+		&v1beta1.PipelineRun{
+			Status: v1beta1.PipelineRunStatus{
+				PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+					PipelineResults: []v1beta1.PipelineRunResult{
+						{Name: "typo_IMAGE_TYPO", Value: *v1beta1.NewStructuredValues("sha256:" + artifactDigest1)},
+						{Name: "typo_IMAGE_URL", Value: *v1beta1.NewStructuredValues(artifactURL1)},
+						{Name: "typo_ARTIFACT_DIGEST", Value: *v1beta1.NewStructuredValues("sha256:" + artifactDigest2)},
+						{Name: "typo_ARTIFACT_TYPO", Value: *v1beta1.NewStructuredValues(artifactURL2)},
+					},
+				},
+			},
+		},
+	)
+
+	prInvalidTypeHintingValue = objects.NewPipelineRunObject(
+		&v1beta1.PipelineRun{
+			Status: v1beta1.PipelineRunStatus{
+				PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+					PipelineResults: []v1beta1.PipelineRunResult{
+						// invalid digest algorithm name
+						{Name: "invalidDigestAlgo_IMAGE_DIGEST", Value: *v1beta1.NewStructuredValues("sha1:" + artifactDigest2)},
+						{Name: "invalidDigestAlgo_IMAGE_URL", Value: *v1beta1.NewStructuredValues(artifactURL2)},
```

Comment:
We may have went too far in the other direction with too many tests 😅 

i.e. what's the difference between these?

```go
{Name: "invalidDigestAlgo_IMAGE_URL", Value: *v1beta1.NewStructuredValues(artifactURL2)},
{Name: "invalidDigestAlgo_ARTIFACT_URI", Value: *v1beta1.NewStructuredValues(artifactURL2)},
{Name: "invalidDigestSHA_IMAGE_URL", Value: *v1beta1.NewStructuredValues(artifactURL2)},
{Name: "invalidDigestSHA_ARTIFACT_URI", Value: *v1beta1.NewStructuredValues(artifactURL2)},
...
```

On the surface they all look like the same test case with different names - can we cut this down so we're only testing each thing once?

Reply:
Sorry for the confusion. Since uri and digest are expected to appear in pair, I tried to cover the case that no subjects are extracted when the digest part is invalid even if the corresponding url in the pair meets the requirement 🤣.